### PR TITLE
Scheduler fixes 

### PR DESCRIPTION
**Pull request by fayebutler**
@currycoder 

- Updated "last run"  to be "last period end".
- Now adds one day to last_period_end and then adds one month
- New errors class, with get, set and send functions to aggregate and send errors for report run to tech-int
- Adding dry run class and argument option to be able to test run through scheduler

### CODE REVIEW

File: scheduler.py [hunk 1]
```diff
@@ -84,24 +89,65 @@ def get_next_run(self, last_run, frequency, frequency_options={}):
...
+            #needs to run at the end of the next period, so add one month
+            next_run = add_one_month(next_run)
+            #make sure set to be correct day 
+            next_run = next_run.replace(day=day) 
             if (now - next_run).days >= 2:
               self.override_data = True
         return next_run
 
 
+class Errors():
```
Comment (currycoder) [inline scheduler.py]: This class should inherit from the `object` builtin

File: scheduler.py [hunk 2]
```diff
@@ -84,24 +89,65 @@ def get_next_run(self, last_run, frequency, frequency_options={}):
...
+    
+    def add_error(self, error):
+        """
+        Add to list of errors
+        """
+        self.errors.append(error)
+        
+        
+    def send_errors(self):
+            """
```
Comment (currycoder) [inline scheduler.py]: Dat indent!

Comment (fayebutler) [inline scheduler.py]: Ooooops.

File: scheduler.py [hunk 3]
```diff
@@ -146,8 +193,43 @@ def _run():
...
                     run_logger.record_run(identifier, run_datetime)
                 except Exception:
-                	traceback.print_exc()
-                	continue
+                    print "Error in generating report %s" % identifier
+                    error_list.add_error("Error in generating report %s : \n %s" % (identifier, traceback.format_exc()))
+                    traceback.print_exc()
+                    continue
+
+def dry_run():
```
Comment (currycoder) [inline scheduler.py]: There's a fair bit of repeated code between this and the `_run` function.  Would it be possible to unify the two with an optional `dry_run` keyword argument?  That way, we know that dry run is definitely a dry run of the exact logic that will be used in the real run

Comment (fayebutler) [inline scheduler.py]: Good point. have done this

**Comment by currycoder** [comment]
Nice work - happy for this to be merged!

**Pull request merged**